flang/lib/Lower/ReductionRuntime.cpp

```diff
@@ -86,6 +86,121 @@ struct ForcedMinvalInteger16 {
   }
 };
 
+/// Placeholder for real*10 version of Product Intrinsic
+struct ForcedProductReal10 {
+  static constexpr const char *name = ExpandAndQuoteKey(RTNAME(ProductReal10));
+  static constexpr Fortran::lower::FuncTypeBuilderFunc getTypeModel() {
+    return [](mlir::MLIRContext *ctx) {
+      auto ty = mlir::FloatType::getF80(ctx);
+      return mlir::FunctionType::get(ctx, {ty}, {ty});
+    };
+  }
+};
+
+/// Placeholder for real*16 version of Product Intrinsic
+struct ForcedProductReal16 {
+  static constexpr const char *name = ExpandAndQuoteKey(RTNAME(ProductReal16));
+  static constexpr Fortran::lower::FuncTypeBuilderFunc getTypeModel() {
+    return [](mlir::MLIRContext *ctx) {
+      auto ty = mlir::FloatType::getF128(ctx);
+      return mlir::FunctionType::get(ctx, {ty}, {ty});
+    };
+  }
+};
+
+/// Placeholder for integer*16 version of Product Intrinsic
+struct ForcedProductInteger16 {
+  static constexpr const char *name =
+      ExpandAndQuoteKey(RTNAME(ProductInteger16));
+  static constexpr Fortran::lower::FuncTypeBuilderFunc getTypeModel() {
+    return [](mlir::MLIRContext *ctx) {
+      auto ty = mlir::IntegerType::get(ctx, 128);
+      return mlir::FunctionType::get(ctx, {ty}, {ty});
+    };
+  }
+};
+
+/// Placeholder for complex(10) version of Product Intrinsic
+struct ForcedProductComplex10 {
+  static constexpr const char *name =
+      ExpandAndQuoteKey(RTNAME(CppProductComplex10));
+  static constexpr Fortran::lower::FuncTypeBuilderFunc getTypeModel() {
+    return [](mlir::MLIRContext *ctx) {
+      auto ty = fir::ComplexType::get(ctx, 10);
+      return mlir::FunctionType::get(ctx, {ty}, {ty});
```

Comment:
Here, the previous comment regarding the `Forcedxxx` applies, but the result argument looks also incorrect to me,  `CppProductComplexXX` return  void and takes the result by ref, so I would expect: `mlir::FunctionType::get(ctx, {}, {refTy, ...})`